service/storage_proxy.cc

```diff
@@ -159,6 +165,85 @@ class storage_proxy::remote {
     netw::connection_drop_slot_t _connection_dropped;
     netw::connection_drop_registration_t _condrop_registration;
 
+    class request_info {
+    public:
+        virtual future<schema_ptr> get_schema() const = 0;
+        virtual dht::token get_token(const schema& schema) const = 0;
+    };
+
+    struct read_request_info: public request_info {
+        schema_ptr _schema;
+        const dht::partition_range& _range;
+    public:
+        read_request_info(schema_ptr schema, const dht::partition_range& range)
+            : _schema(schema)
+            , _range(range)
+        {
+        }
+        virtual future<schema_ptr> get_schema() const override {
+            return make_ready_future<schema_ptr>(_schema);
+        }
+        virtual dht::token get_token(const schema&) const override {
+            return end_token(_range);
+        }
+    };
+
+    class write_request_info: public request_info {
+        remote& _remote;
+        table_schema_version _schema_version;
+        netw::msg_addr _msg_from;
+        partition_key_view _partition_key;
+    public:
+        write_request_info(remote& remote, table_schema_version schema_version, netw::msg_addr msg_from,
+            partition_key_view _partition_key)
+            : _remote(remote)
+            , _schema_version(schema_version)
+            , _msg_from(msg_from)
+            , _partition_key(std::move(_partition_key))
+        {
+        }
+        virtual future<schema_ptr> get_schema() const override {
+            return _remote.get_schema_for_write(_schema_version, _msg_from);
+        }
+        virtual dht::token get_token(const schema& schema) const override {
+            return schema.get_partitioner().get_token(schema, _partition_key);
+        }
+    };
```

Comment:
So in this version instead of passing a function to retrieve the info you provided even more elaborate virtual interface with two virtual function calls, but I have the same comment. Why not pass the token directly and  not call  `get_schema_for_write` or `get_schema_for_read` from try_apply_fence directly (we can have read and write version of the function)? Why dispatch in runtime something that can be dispatched in compile time?

Reply:
There are difference interfaces for read and write code paths's, they are expressed now in different implementations of `get_schema` and `get_token`. Abstraction is a way to remove duplication, which we would also have to deal with if we had separate functions `get_schema_for_write` and `get_schema_for_read`. We could turn this into a template to get rid of runtime dispatching, but I don't see the value in it - it's more complicated, incurs code bloat, and vcall is not on a hot path here.

Reply:
Not all differences has to be encapsulated into a virtual class. Especially if the differences are apparent in compile time. `get_token` can be replaced by simply passing tokens to the function. Each caller knows how to retrieve it. calling of `get_schema_for_write`/`get_schema_for_read` can be controlled by simple bool flag if you have a dislike for templates.

Reply:
If there is so much differences between read and write fencing then it is better to have two separate functions. They can still call a common one after acquiring  a schema and a token.

Reply:
All the differences are only in parameters, the different ways of getting schema and token. The basic logic is the same - loop, check version, wait, token ownership check. We would have either copy-paste or clunky helper functions without abstracting those differences away. I really don't get why you don't like it.

Reply:
Ownership check can be moved into a generic function (after acquiring the schema and the tokens) . Loop will be repeated, yes, but there is a also a lot of repeated boilerplate code in virtual interface impl. But we will have compile time dispatch instead.

Reply:
> But we will have compile time dispatch instead.

Why it's important? What value compile time dispatch can bring here?

> Loop will be repeated, yes, but there is a also a lot of repeated boilerplate code in virtual interface impl.

The virtual interface impl boilerplate is standard for OOP and trivial (meaning easy to maintain), copy-pasted loop is not.

> Ownership check can be moved into a generic function (after acquiring the schema and the tokens)

It would need to have strange parameters `fence.topology_version, current_version` to pass them to `slogger.trace` and `stale_topology_exception`.

Reply:
> > But we will have compile time dispatch instead.
> 
> Why it's important? What value compile time dispatch can bring here?

Simplicity and efficiency. Jumping through virtual functions is relatively expensive.

> 
> > Loop will be repeated, yes, but there is a also a lot of repeated boilerplate code in virtual interface impl.
> 
> The virtual interface impl boilerplate is standard for OOP and trivial (meaning easy to maintain), copy-pasted loop is not.

OOP is not in favor among non Java programmers lately. Procedural programing is fine if it does the job. To be OOP is not a goal.

> 
> > Ownership check can be moved into a generic function (after acquiring the schema and the tokens)
> 
> It would need to have strange parameters `fence.topology_version, current_version` to pass them to `slogger.trace` and `stale_topology_exception`.

Internal function may take whatever parameters it needs to. It is an internal function.

Reply:
> Simplicity and efficiency. Jumping through virtual functions is relatively expensive.

Efficiency doesn't matter here, it's not a hot path. And these are just three simple classes.

> OOP is not in favor among non Java programmers lately. Procedural programing is fine if it does the job. To be OOP is not a goal.

These are not practical statements. C++ is OOP language, among other paradigms, and it can be used if it's the right tool for the job.

This discussion has clearly turned to the matters of taste and bikeshedding.

Reply:
I agree it's enough. Usually in such case it's best to appeal to the second opinion, and side on the side of the majority.

Reply:
> > Simplicity and efficiency. Jumping through virtual functions is relatively expensive.
> 
> Efficiency doesn't matter here, it's not a hot path. And these are just three simple classes.

It does not mean we should insert vtable just because we can.

> 
> > OOP is not in favor among non Java programmers lately. Procedural programing is fine if it does the job. To be OOP is not a goal.
> 
> These are not practical statements. C++ is OOP language, among other paradigms, and it can be used if it's the right tool for the job.

Absolutely can. Here it is not the right tool for the job. See below. 

> 
> This discussion has clearly turned to the matters of taste and bikeshedding.

It is not. I have two problems with the approach:

First the code  erases the information it has at the call site just to get it back through vtables later. There simply no reason for that. This is needed when the same call site may produce different types depending on some other input which is clearly not the case here, so virtual interface is just abused here.

Second  it creates ad-hoc interface which just make the code  harder to read. When you look at the `try_apply_fence` implementation you see calls to `get_schema()` and `get_tokens()` and you need to figure out what each of them is doing. Without virtual interface it is immediately obvious. You objection for having two functions that makes it clear is 3 line of code duplication, but when I suggest to use a template to avoid code duplication the objection is general dislike of templates?

Reply:
> First the code erases the information it has at the call site just to get it back through vtables later.

This is just how virtual functions work in general. At some place in the code you are making decision about some strategy (choose the concrete impl), erase all the details, and later use only the abstract interface. There
are two call sites to `try_apply_fence` and by erasing the differences between them we can treat them uniformly.

> Second it creates ad-hoc interface which just make the code harder to read. When you look at the try_apply_fence implementation you see calls to get_schema() and get_tokens() and you need to figure out what each of them is doing. Without virtual interface it is immediately obvious.

This is a vague argument, it can be said about any use of abstractions. Of course they hide some information to provide a uniform interface. It's just a matter of whether the abstraction pays off or not, whether it's value outweighs it's cognitive burden or not. And that depends a lot on taste and habit.

> You objection for having two functions that makes it clear is 3 line of code duplication, but when I suggest to use a template to avoid code duplication the objection is general dislike of templates?

I consider templates as 'more advanced' abstraction tool compared to virtual functions. They are usually more difficult to reason about, and they cause a code bloat.

I updated the code to use `std::variant`, this seems to solve our disagreement.